sql/core/src/main/scala/org/apache/spark/sql/execution/SparkOptimizer.scala

```diff
@@ -37,7 +37,11 @@ class SparkOptimizer(
 
   override def earlyScanPushDownRules: Seq[Rule[LogicalPlan]] =
     // TODO: move SchemaPruning into catalyst
-    SchemaPruning :: V2ScanRelationPushDown :: V2Writes :: PruneFileSourcePartitions :: Nil
+    SchemaPruning :: RowLevelCommandScanRelationPushDown :: V2ScanRelationPushDown :: V2Writes ::
```

Comment:
Is it possible to merge `RowLevelCommandScanRelationPushDown` into `V2ScanRelationPushDown` so they're in one place?

Reply:
Keeping them separate for now as `V2ScanRelationPushDown` is already complicated and none of that logic applies.

Reply:
After looking at this more, I agree with this direction. There is no need to overcomplicate either case.